bnd run prevents use of JLine Gogo shell

I'm trying to use the `org.apache.felix.gogo.jline` bundle to get command history, highlighting etc in Gogo shell. This works perfectly if I export my bndrun file and run the resulting JAR with `java -jar application.jar`. However when I run the bndrun file directly with `bnd run application.bndrun` then bnd does something to the terminal which prevents JLine from using it:

    Dec 19, 2017 4:15:48 PM org.jline.utils.Log logr
    WARNING: Unable to create a system terminal, creating a dumb terminal (enable debug logging for more information)

After turning on debug logging (i.e. edit `$JAVA_HOME/jre/lib/logging.properties`, set default level=FINEST):

    Dec 19, 2017 4:16:40 PM org.jline.utils.Log logr
    WARNING: Creating a dumb terminal
    java.lang.IllegalStateException: Unable to create a system terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:184)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:149)
        at org.apache.felix.gogo.jline.Activator.startShell(Activator.java:117)
        at org.apache.felix.gogo.jline.Activator.serviceFound(Activator.java:73)
        at org.apache.felix.gogo.jline.SingleServiceTracker.update(SingleServiceTracker.java:168)
        at org.apache.felix.gogo.jline.SingleServiceTracker.findMatchingReference(SingleServiceTracker.java:124)
        at org.apache.felix.gogo.jline.SingleServiceTracker.open(SingleServiceTracker.java:90)
        at org.apache.felix.gogo.jline.Activator.start(Activator.java:57)
        at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697)
        at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2146)
        at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
        at aQute.launcher.Launcher.startBundles(Launcher.java:502)
        at aQute.launcher.Launcher.activate(Launcher.java:409)
        at aQute.launcher.Launcher.run(Launcher.java:289)
        at aQute.launcher.Launcher.main(Launcher.java:142)
        Suppressed: java.lang.NoClassDefFoundError: com/sun/jna/Platform
            at org.jline.terminal.impl.jna.JnaNativePty.current(JnaNativePty.java:37)
            at org.jline.terminal.impl.jna.JnaSupportImpl.current(JnaSupportImpl.java:15)
            at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:217)
            ... 15 more
        Caused by: java.lang.ClassNotFoundException: com.sun.jna.Platform not found by org.jline [11]
            at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1639)
            at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
            at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
            at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
            ... 18 more
        Suppressed: java.io.IOException: Not a tty
            at org.jline.terminal.impl.ExecPty.current(ExecPty.java:44)
            at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:226)
            ... 15 more
        Caused by: java.io.IOException: Error executing 'tty': not a tty
            at org.jline.utils.ExecHelper.exec(ExecHelper.java:42)
            at org.jline.terminal.impl.ExecPty.current(ExecPty.java:41)
            ... 16 more

--- a/aQute.libg/src/aQute/libg/command/Command.java
+++ b/aQute.libg/src/aQute/libg/command/Command.java
@@ -108,6 +108,8 @@ public class Command {
 		}
 
 		p.directory(cwd);
+		if (in == System.in)
+			p.redirectInput(ProcessBuilder.Redirect.INHERIT);
 		process = p.start();
 
 		// Make sure the command will not linger when we go
@@ -174,7 +176,6 @@ public class Command {
 					rdInThread.setDaemon(true);
 					rdInThread.start();
 				} else {
-
 					IO.copy(in, stdin);
 					stdin.close();
 				}
@@ -193,7 +194,8 @@ public class Command {
 		byte exitValue = (byte) process.waitFor();
 		finished.set(true);
 		if (rdInThread != null) {
-			IO.close(in);
+			if (in != System.in)
+				IO.close(in);
 			rdInThread.interrupt();
 		}
 

